ext/TensorOperationsChainRulesCoreExt.jl

```diff
@@ -18,7 +18,21 @@ trivtuple(N) = ntuple(identity, N)
 @non_differentiable TensorOperations.tensorcontract_structure(args...)
 @non_differentiable TensorOperations.tensorcontract_type(args...)
 @non_differentiable TensorOperations.tensoralloc_contract(args...)
-@non_differentiable TensorOperations.tensorfree!(args...)
+
+# Cannot free intermediate tensors when using AD!
+function ChainRulesCore.rrule(::typeof(TensorOperations.tensorfree!), args...)
+    tensorfree!_pullback(Δargs...) = ntuple(x -> NoTangent(), length(args))
+    return nothing, tensorfree!_pullback
+end
+function ChainRulesCore.rrule(::typeof(TensorOperations.tensoralloc), ttype, structure,
+                              istemp, backend...)
+    output = TensorOperations.tensoralloc(ttype, structure, false, backend...)
+    function tensoralloc_pullback(Δoutput)
+        istemp && TensorOperations.tensorfree!(unthunk(Δoutput), backend...)
+        return ntuple(x -> NoTangent(), 4 + length(backend))
+    end
+    return output, tensoralloc_pullback
```

Comment:
I am confused by this one. You do not want to free a temporary tensor in the context of AD. So I understand that you rewrite the forward function for `tensorfree!` to do nothing. But I assume you do then want to free it after AD has passed through. So I would think that `tensoralloc_pullback` frees the output of `tensoralloc`, i.e. the variable `output` (not `Δoutput`), and since it is manually freed, that it is still fine to allocate `output` using the original `istemp` value, without it needing to be replaced by `false`?

Reply:
I'll have to think about this a bit more tomorrow actually, because it's not even that simple I think:

I'm not sure how safe it is to allow any temporary allocation and freeing in the context of AD, I don't know if Zygote guarantees it will even call that rule, if it can show that it does not use the gradient. I also don't know how many times nor in what order these functions get executed
 It's definitely too much of a magic black box to me to be able to confidently say my current implementation is right, and we might be best off just not performing any management of the temporaries.

In any case, I could try this with the manual malloc and free, and try and break it 😀 this should immediately SegFault so that might be worth some experimenting.

Reply:
Not sure if, even if you free a pointer, whether it will immediately segfault if you try to use it. This works just fine:
```julia
a = Libc.malloc(sizeof(Float64) * 32);
b = unsafe_wrap(Array, convert(Ptr{Float64}, a), 32; own = false)
copyto!(b, randn(32))
Libc.free(a)
b[1] = 0.5
```
However, calling `norm(b)` does segfault. I do not know how these things work exactly.

Reply:
But I agree, we can probably not rely on the backward rules being called, so it is probably better to disable the `temp` argument in `tensoralloc` and the forward action of `tensorfree` and leave it at that.